src/vos/vos_query.c

```diff
@@ -255,113 +222,282 @@ _query_recx(struct open_query *query, daos_recx_t *recx, struct evt_extent *filt
 	return rc;
 }
 
-static int
-query_normal_recx(struct open_query *query, daos_recx_t *recx)
+typedef bool (*check_func)(const struct evt_entry *ent1, const struct evt_entry *ent2);
+
+static bool
+hi_gtr(const struct evt_entry *ent1, const struct evt_entry *ent2)
 {
-	struct evt_extent filter_ex;
-	int		  opc;
+	return ent1->en_sel_ext.ex_hi > ent2->en_sel_ext.ex_hi;
+}
 
-	memset(recx, 0, sizeof(*recx));
-	/* query visible last recx */
-	opc = EVT_ITER_EMBEDDED | EVT_ITER_VISIBLE | EVT_ITER_SKIP_HOLES;
-	if (query->qt_flags & VOS_GET_MAX)
-		opc |= EVT_ITER_REVERSE;
+static bool
+lo_lt(const struct evt_entry *ent1, const struct evt_entry *ent2)
+{
+	return ent1->en_sel_ext.ex_lo < ent2->en_sel_ext.ex_lo;
+}
 
-	filter_ex.ex_lo = 0;
-	filter_ex.ex_hi = ~(uint64_t)0;
+static bool
+is_after(const struct evt_entry *ent1, const struct evt_entry *ent2)
+{
+	return ent1->en_sel_ext.ex_lo > ent2->en_sel_ext.ex_hi;
+}
 
-	return _query_recx(query, recx, &filter_ex, opc);
+static bool
+is_before(const struct evt_entry *ent1, const struct evt_entry *ent2)
+{
+	return ent1->en_sel_ext.ex_hi < ent2->en_sel_ext.ex_lo;
 }
 
-static int
-query_ec_normal_recx(struct open_query *query, daos_recx_t *recx,
-		     struct evt_extent *filter_ex, bool skip_hole)
+static void
+overlap_lo(struct evt_entry *nentry, struct evt_entry *pentry, bool *nrefresh, bool *prefresh)
 {
-	struct evt_extent tmp_ex;
-	int		  opc;
+	if (nentry->en_sel_ext.ex_lo == pentry->en_sel_ext.ex_lo) {
```

Comment:
looks like it can be  just <=  and remove the following if, since nent is hole, you just fetch next both p and n to compare?

Reply:
No, because the nentry is a hole.

Let's say stripe size is 16 and nentry is punch from 0-31 and parity is from 16-31.  If you fetch the next nentry, you'll miss the fact that the next parity may be punched.